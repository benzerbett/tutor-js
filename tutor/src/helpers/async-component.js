import { asyncComponent as asyncLoader } from 'react-async-component';
import ErrorComponent from '../components/error-monitoring/async-load-error';
import LoadingComponent from '../components/loading-screen';

export function asyncComponent(resolve) {
  return asyncLoader({
    resolve,
    LoadingComponent,
    ErrorComponent,
  });
}

export function loadAsync(resolve) {
  // return a function so the router will only evaluate it when it's needed
  // in the future we can insert role dependant logic here
  return () => asyncComponent(resolve);
}
