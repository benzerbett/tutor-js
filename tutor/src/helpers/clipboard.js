import _ from 'underscore';

const SUPPORTED = typeof document.queryCommandSupported === 'function' ? document.queryCommandSupported('copy') : undefined;

export default {
  isSupported: _.constant(!!SUPPORTED),

  copy() {
    if (!SUPPORTED) { return; }

    try {
      return document.execCommand('copy');
    } catch (e) {
      return console.warn('clipboard copy failed', e);
    }
  },
};
