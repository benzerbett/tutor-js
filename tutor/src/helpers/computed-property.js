import { forIn } from 'lodash';
import { computed } from 'mobx';

export function addComputedProperty(obj, name, fn) {

  Object.defineProperty(obj, name, {
    configurable: true,
    get() {
      const getter = computed(fn, { context: this });
      Object.defineProperty(this, name, {
        configurable: false,
        get: () => getter.get(),
      });
      return getter.get();
    },
  });

}

export function extendHasMany(properties) {
  return (collection) => {
    forIn(properties, (fn, prop) => {
      addComputedProperty(collection, prop, fn);
    });
  };
}
