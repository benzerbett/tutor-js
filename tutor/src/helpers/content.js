import _ from 'underscore';


export default {
  // Convert each number to base 10 with it's position based on index.
  // If section is not present, 0 is set for it
  //   1 becomes 100, 1.1 becomes 101, 3.2 to 302, 3.2.1 -> 30201
  // Useful for sorting
  chapterSectionToNumber(chapter_section) {
    if (_.isString(chapter_section)) { chapter_section = chapter_section.split('.'); }
    if (chapter_section.length === 1) { chapter_section.push(0); } // add a section 0 if it has only a chapter
    let position = -1;
    return _.reduceRight(chapter_section, (memo, num) => memo + (num * Math.pow(100, (position += 1)))
      , 0);
  },
};
