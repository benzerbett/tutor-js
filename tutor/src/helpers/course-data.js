import Courses from '../models/courses-map';

import Router from '../helpers/router';
import TimeHelper from '../helpers/time';

export default {
  getCourseDataProps(courseId) {
    let dataProps;
    if (courseId == null) {
      ((((((({ courseId } = Router.currentParams())))))));
    }
    const course = Courses.get(courseId);
    return dataProps = {
      'data-title': course.name,
      'data-book-title': course.bookName || '',
      'data-appearance': course.appearance_code,
    };
  },

  getCourseBounds(courseId) {
    const course = Courses.get(courseId);

    const start = TimeHelper.getMomentPreserveDate(course.starts_at, TimeHelper.ISO_DATE_FORMAT);
    const end = TimeHelper.getMomentPreserveDate(course.ends_at, TimeHelper.ISO_DATE_FORMAT);

    return { start, end };
  },
};
