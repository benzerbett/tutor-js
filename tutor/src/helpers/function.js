export default {
  // coffeelint: disable=no_empty_functions
  empty() {},
};
// coffeelint: enable=no_empty_functions
