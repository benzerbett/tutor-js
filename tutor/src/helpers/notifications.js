import { NotificationActions } from 'shared';

export default {
  start(bootstrapData) {
    NotificationActions.startPolling();
    return (() => {
      const result = [];
      const object = bootstrapData.flash || {};
      for (let level in object) {
        const message = object[level];
        result.push(NotificationActions.display({ message, level }));
      }
      return result;
    })();
  },

  buildCallbackHandlers(comp) {
    // Require router here because requiring it in global scope
    // causes hot module reloading to fail in dev mode because this file is required outside the HMR scope
    const Router = require('./router');
    const { router } = comp.context;
    if (!router) {
      throw new Error('Component\'s context must have router present');
    }
    return {
      missing_student_id: {
        onAdd({ course }) {
          return router.history.push(
            Router.makePathname('changeStudentId', { courseId: course.id })
          );
        },
      },
    };
  },
};
