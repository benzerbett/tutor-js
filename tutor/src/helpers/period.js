import { filter, isEmpty, sortBy, toNumber, isNumber } from 'lodash';
import { addComputedProperty } from './computed-property';
import S from './string';

// Used to filter periods by helper methods
const isArchivedCheckFn = period => period.is_archived;
const isActiveCheckFn   = period => !period.is_archived;

const NumberLike = /[^0-9]+/;


const PeriodHelper = {
  getOrder(period) {
    return S.getNumberAndStringOrder(period.name);
  },

  sort(periods) {
    // expects either numbers, names with numbers or just names
    periods = sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const name = period.name.match(/[^0-9]+/ig);
        if (name) {
          return name;
        }
      }
    });
    return sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const number = period.name.match(/[0-9.-]+/g);
        if (number) {
          return parseFloat(number);
        }
      } else {
        return period.name;
      }
    });
  },

};


export default PeriodHelper;
