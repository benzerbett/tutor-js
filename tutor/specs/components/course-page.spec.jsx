import { React } from './helpers/component-testing';
import Renderer from 'react-test-renderer';
import COURSE from '../../api/courses/1.json';
import Course from '../../src/models/course';
import CoursePage from '../../src/components/course-page';

describe('Course Page', () => {

  let props;

  beforeEach(function() {
    props = {
      course: new Course(COURSE),
      className: 'test-page',
    };
  });

  const Title = () => <h1>A Big Bad Page</h1>;

  it('renders and matches snapshot', () => {
    expect(Renderer.create(
      <CoursePage {...props}>
        <h3>Hello, this is the body</h3>
      </CoursePage>
    ).toJSON()).toMatchSnapshot();
    expect(Renderer.create(
      <CoursePage
        {...props}
        title={<Title />}
        notices={<h1>This is a test notice</h1>}
        subtitle="This is a subtitle"
        controls={<button>Click Me</button>}
      >
        <h3>Hello, this is the body</h3>
      </CoursePage>
    ).toJSON()).toMatchSnapshot();});
});
