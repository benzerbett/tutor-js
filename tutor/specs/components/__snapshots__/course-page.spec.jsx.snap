// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course Page renders and matches snapshot 1`] = `
<div
  className="course-page test-page"
  data-appearance="college_physics"
  data-book-title="College Physics"
  data-is-preview={false}
  data-term="Spring 2017"
  data-title="College Physics"
>
  <header />
  <div
    className="body-wrapper"
  >
    <div
      className="body"
    >
      <h3>
        Hello, this is the body
      </h3>
    </div>
  </div>
</div>
`;

exports[`Course Page renders and matches snapshot 2`] = `
<div
  className="course-page test-page"
  data-appearance="college_physics"
  data-book-title="College Physics"
  data-is-preview={false}
  data-term="Spring 2017"
  data-title="College Physics"
>
  <header>
    <div
      className="title-wrapper"
    >
      <div
        className="title"
      >
        <h1>
          A Big Bad Page
        </h1>
      </div>
      <div
        className="subtitle"
      >
        This is a subtitle
      </div>
    </div>
    <h1>
      This is a test notice
    </h1>
    <div
      className="controls-wrapper"
    >
      <button>
        Click Me
      </button>
    </div>
  </header>
  <div
    className="body-wrapper"
  >
    <div
      className="body"
    >
      <h3>
        Hello, this is the body
      </h3>
    </div>
  </div>
</div>
`;
